Asgard/Communications/Classes/GridConnectProcessor.cs

```diff
@@ -0,0 +1,166 @@
+            ReadPipe(pipe.Reader).AsgardFireAndForget();
+            Listen(pipe.Writer).AsgardFireAndForget();
+        }
+
+        private Task Listen(PipeWriter writer)
+        {
+            return Task.Run(async () => {
+                const int minBufferSize = 64;
+                while (!cts.IsCancellationRequested)
+                {
+                    var memory = writer.GetMemory(minBufferSize);
+                    try
+                    {
+                        var read = await Transport.ReadAsync(memory, cts.Token);
+                        if (read == 0)
+                        {
+                            //todo?
+                        }
+
+                        writer.Advance(read);
+                        await writer.FlushAsync(cts.Token);
+                    }
+                    catch (TaskCanceledException)
+                    {
+                        //ok
+                    }
+                    catch (Exception e)
+                    {
+                        TransportError?.Invoke(this, new TransportErrorEventArgs(new TransportException("Error receiving bytes", e)));
+                    }
+                }
+            });
+        }
+
+        private Task ReadPipe(PipeReader reader)
+        {
+            return Task.Run(async () => {
+                while (!cts.IsCancellationRequested)
+                {
+                    var result = await reader.ReadAsync(cts.Token);
+                    var buffer = result.Buffer;
+                    SequencePosition? endPosition;
+                    do
+                    {
+                        endPosition = buffer.PositionOf((byte)';');
+                        if (endPosition != null)
+                        {
+                            endPosition = buffer.GetPosition(1, endPosition.Value);
+                            ProcessMessage(buffer.Slice(0, endPosition.Value));
+                            buffer = buffer.Slice(endPosition.Value);
+                        }
+
+                    } while (endPosition != null);
+
+                    reader.AdvanceTo(buffer.Start, buffer.End);
+                }
+            });
+        }
+
+        private void ProcessMessage(ReadOnlySequence<byte> readOnlySequence)
+        {
+            var startPosition = readOnlySequence.PositionOf((byte)':');
+            if (startPosition != null)
+            {
+                var msg = GetMessageString(readOnlySequence.Slice(startPosition.Value, readOnlySequence.End));
+                logger?.LogTrace("Message received {0}", msg);
+                GridConnectMessage?.Invoke(this, new MessageReceivedEventArgs(msg));
+            }
+            else
+            {
+                //TODO: is there a way to sensibly combine the two string generations to only do it once, but also only when needed
+                logger?.LogWarning("Partial message received: {0}", GetMessageString(readOnlySequence));
```

Comment:
Use a Lazy<string>
var messageString = new Lazy<string>(() => GetMessageString(readOnlySequence));
then in both logger? and TransportError?
messageString.Value